designs/_posts/2016-09-30-repository-cache.md

```diff
@@ -0,0 +1,91 @@
+---
+layout: contribute
+title: Central cache for external repositories
+---
+
+# Central cache for external repositories
+## Part 1: Caching HttpDownloader values
+
+**Author**: jingwen@google.com
+
+**Status**: Implemented
+
+**Reviewers**: kchodorow@, jcater@, dmarting@
+
+## Introduction
+
+With `bazel fetch //repo:...`, external repositories and artifacts are
+downloaded from remote sources with repository functions, such as
+`http_archive()` or `maven_jar()`. Sometimes, network access is not
+available, or redownloading of already-downloaded repositories is
+wasteful.
+
+With a central cache, Bazel can check it for the requested
+repositories before hitting the internet. This allows multiple
+workspaces to share the same repository cache. The cache should also
+be portable, i.e. tar’ing up the cache and reusing it on another
+system.
+
+### Related issues:
+
+Make `maven_jar` and friends smarter by re-using previously fetched
+artifacts across different projects
+[#1752](https://github.com/bazelbuild/bazel/issues/1752).
+
+## Objective
+
+Implement an additional step prior to fetching that checks a directory
+`$repository_cache` for the requested external repository. 
+
+Note:
+
+- This does not change the way `output_base/external` is organized. It
+is delegating the actual location of the downloaded repositories to
+somewhere outside of `output_base/external` (which means outside of
+individual Bazel invocation folders), and symlinking to those
+locations instead.
+- This does not require a change in CI configuration until it’s
+switched on by default. It can be used with the flag 
+`--experimental_repository_cache` during the development process.
+
+## Not in scope
+
+- (Skylark) Reconciling caching strategies between Bazel and other tools
+like Maven, which has their own strategies. It’s possible to cache the
+output of `repository_ctx.execute`, while still respecting the external
+tool’s caching strategy, but that is not in the scope of this design.
+- Human-readable cache layouts. Simplifies inspection of cache content.
+- Garbage collection / cache cleanup of non-referenced cache content.
+- Using a remote filesystem as cache (NFS, ssh-fuse, etc)
```

Comment:
nit: We usually avoid using "etc." since it leaves the sentence trailing off. Use "and so on" with a period instead.

Reply:
Done.